### PR TITLE
[SPARK-51182][SQL] DataFrameWriter should throw dataPathNotSpecifiedError when path is not specified

### PR DESCRIPTION
**Pull request by vrozov**
### What changes were proposed in this pull request?
Change error message from `Expected exactly one path to be specified, but got: .` to `'path' is not specified.` when path is not specified in the call to `DataFrame.write().save(path)` explicitly or using `option(path, ...)`, `parquet(path)` and etc.

### Why are the changes needed?
The error message is more accurate. 

### Does this PR introduce _any_ user-facing change?
Yes, user would get corrected error message when they do not specify path.

### How was this patch tested?
Updated error message in the R test suite.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by vrozov** [comment]
@HeartSaVioR Please review.

**Comment by HeartSaVioR** [comment]
I'll defer the review to SQL area committers, cc. @cloud-fan

**Comment by vrozov** [comment]
@cloud-fan Please review

**Comment by vrozov** [comment]
@HeartSaVioR can you please suggest another committer to review this small PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -462,8 +462,10 @@ case class DataSource(
     val allPaths = paths ++ caseInsensitiveOptions.get("path")
     val outputPath = if (allPaths.length == 1) {
       makeQualified(new Path(allPaths.head))
-    } else {
+    } else if (allPaths.length > 1) {
       throw QueryExecutionErrors.multiplePathsSpecifiedError(allPaths)
+    } else {
+      throw QueryExecutionErrors.dataPathNotSpecifiedError()
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:468]: Do you know if there's any test case that can cover the new branch? If not, could you add one in this pr? @vrozov

**Comment by cloud-fan** [comment]
The change LGTM, can we add a test in `DataFrameReaderWriterSuite`?

**Comment by vrozov** [comment]
@cloud-fan Added tests, please check.

**Review by cloud-fan — APPROVED** [review]

**Comment by vrozov** [comment]
@cloud-fan any action on my side before the PR is merged?

**Comment by cloud-fan** [comment]
@vrozov can you remove the java test? https://github.com/apache/spark/pull/49928#discussion_r1981021185

**Review by vrozov — COMMENTED** [review]
@cloud-fan please see my response.

**Comment by LuciferYang** [comment]
> @vrozov can you remove the java test? [#49928 (comment)](https://github.com/apache/spark/pull/49928#discussion_r1981021185)

+1, for @cloud-fan's comments

**Comment by vrozov** [comment]
@cloud-fan @LuciferYang I prefer to keep the test in the java as it does not hurt and 

1. There is similar test in R even though it is not R specific
2. Other tests in `JavaDataFrameReaderWriterSuite.java` are not java specific either

**Comment by cloud-fan** [comment]
We should not encourage people to add Java tests for non-Java-related stuff. It doesn't hurt but it's about setting a consistent dev guideline for all contributors.

**Comment by vrozov** [comment]
@cloud-fan I agree that non language specific tests should be discouraged.  The best option to do this is to remove the test case from R and remove java files such as `JavaDataFrameReaderWriterSuite.java`. With such files present contributors will still be confused and add test cases to **existing** java test suites.

**Comment by cloud-fan** [comment]
We need basic functionality tests in different languages, but regression tests shouldn't go there if it's not language-specific.

**Comment by cloud-fan** [comment]
To put it simple, the Java test suites should be created with all basic functionalities at the beginning, and we only add new tests to them if they are Java-specific.

It may still be fuzzy but for this case which improves the error message, it happens a lot and we shouldn't require other PR authors to add Java tests for such changes.

**Comment by cloud-fan** [comment]
@vrozov if you really want to improve the Java test coverage of `.save`, that's great, but let's do it right. We need both positive and negative test cases, not only target one corner case of path not specified.

**Comment by vrozov** [comment]
@cloud-fan The positive test cases are already covered in the `JavaDataFrameReaderWriterSuite.java`. Please see https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L82, https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L92, https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L144, https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L135, 
https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L126,
https://github.com/apache/spark/blob/7019d5e63b7218049bacf3392716bf6faf8f82a1/sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameReaderWriterSuite.java#L153
and if there are more positive basic use cases that are not covered, adding them should be part of a separate PR (as this one handles specific negative case for the basic API).

Can you please also confirm 
>Was the approach to testing discussed on dev list and is documented in the contribution guidelines?

**Comment by cloud-fan** [comment]
We don't document it, but we have been doing it for many years. Please check all the SQL bug fixes, none of them added Java test unless the bug is Java related.

**Comment by vrozov** [comment]
> We don't document it, but we have been doing it for many years.

IMO, it needs to be discussed on dev list if it was not discussed already (at least formally) and documented in the contribution guidelines. Without the documentation many new contributors would be in the dark. It is not practical to expect that new contributors will check all the SQL bug fixes.

**Comment by cloud-fan** [comment]
It's pretty natural not to add tests in all the languages for bug fixes that are not language-specific, as you would need to add Scala, Java, Python, and R. With Spark Connect, Spark supports more languages like Go and Rust. It's impractical to ask Spark contributors to write tests in so many languages.

I don't understand why you pushed so hard for the Java test. We appreciate your contribution, but it's really a waste of time on a debate like this. Anyway, feel free to raise this to the dev list if you want to get more opinions to reach a consensus.

**Comment by vrozov** [comment]
@cloud-fan There are two separate issues discussed on the PR.

1. Approach to the testing functionality in different languages (Scala, Java, Python, R, Go). I am fine with whatever approach was agreed by the community. My only ask is to document the approach if it was not documented. At ASF, community takes the priority over code, so if there are any code convention (style, testing, etc), they need to be discussed and documented for the benefit of new community members.

2. Should the test case for `save()` be part of languages other than Scala. IMO, yes, this is why it is part of `R` and it was actually a miss in Scala and Java.

**Comment by cloud-fan** [comment]
OK, given that ".save without path" is tested in R, for consistency we should also test it in Scala and Java. Can you also check `python/pyspark/sql/tests/test_readwriter.py` and add the test case if it's missing? It's ok if you are not familiar with Python and we can leave it as a follow-up work.

**Comment by vrozov** [comment]
Added to `ReadwriterTestsMixin`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
